akka-actor/src/main/scala/akka/actor/dungeon/DeathWatch.scala

```diff
@@ -81,15 +81,16 @@ private[akka] trait DeathWatch { this: ActorCell =>
           watching = removeFromMap(actor, watching)
         }
         if (!isTerminating) {
-          self.tell(optionalMessage.getOrElse(Terminated(actor)(existenceConfirmed, addressTerminated)), actor)
-          terminatedQueuedFor(actor)
+          self.tell(Terminated(actor)(existenceConfirmed, addressTerminated), actor)
+          terminatedQueuedFor(actor, optionalMessage)
         }
     }
     if (childrenRefs.getByRef(actor).isDefined) handleChildTerminated(actor)
   }
 
-  private[akka] def terminatedQueuedFor(subject: ActorRef): Unit =
-    terminatedQueued += subject
+  private[akka] def terminatedQueuedFor(subject: ActorRef, customMessage: Option[Any]): Unit =
+    if (!terminatedQueued.contains(subject))
```

Comment:
This check is to ensure that the `terminatedQueuedFor` invoked in `Stash#enqueueFirst` doesn't replace a potential custom message.